### ISSUE TITLE
Support for S3 private buckets

### ISSUE THREAD

**Issue by g2giovanni** [radiantearth/stac-browser#83]
Hello everyone,
is there a way to use stac-browser with COG files stored on a private S3 bucket?

**Comment by m-mohr** [comment]
I don't think so, but I've never worked with private S3 buckets so I can't really tell, unfortunately.

**Comment by rowanwins** [comment]
Ultimately I suspect this is a problem that will need to be solved outside of the stac-browser client app. 

If we limit the problem to the rendering of COGS then I can see two main approaches
1. **Use presigned-urls.** 
Currently when the client requests tiles it sends a url like `https://myendpoint/cog/tiles/1/2/3?url=https://somewhere.com/mycog.tif`. As it stands the Tiling Infrastructure (TiTiler or RadiantTiler) doesn't have access to that COG file. However [presigned urls](https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html) are  a way of providing access to secure file via a URL. So the request to the tiler might look something like  `https://myendpoint/cog/tiles/1/2/3?url=https://somewhere.com/mycog.tif?a-bunch-of-aws-presigned-params`.  
You would however need a server with the bucket credentials that could generate the pre-signed url for you, and we'd also need to look at whether TiTiler and RadiantTiler would respect those url parameters. Maybe within the app there could be a config flag for the app that specified an endpoint for retrieving presigned urls, but we leave it up to each deployer to write their own server for generating those urls?

2. **Enhance TiTiler or RadiantTiler to better support auth**
 Somehow write the AWS auth process into TiTiler or the RadiantTiler.

Hope that helps

**Comment by m-mohr** [comment]
Yes, I suspect the same.

One note regarding the tiling: STAC Browser also supports client side COGs and TiTiler or RadiantTiles might not even be involved.

**Comment by DanielJDufour** [comment]
Hi, all 👋 

I'm a big fan of the pre-signed URLs that @rowanwins mentioned.  They definitely make things easier with the client-side visualization libraries.  I think it's important to note that (last I checked), geotiff.js doesn't provide an easy way to provide authentication headers to the internal metadata and tile requests.  We could potentially add it through a PR to geotiff.js.

I think TiTiler (because it uses rasterio/gdal) already supports reading from private or requester pays S3 buckets.  You just have to set the AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY environmental variables.  I think I tested this a year or so ago accessing NAIP requester pays imagery, but can't recall 100%.  Of course with the caveat that this is for buckets that your S3 account has access to.  I'm honestly not sure what to do if an external user is "requesting" access to a bucket owned by someone else.

I like @rowanwins idea of having an authentication/signing endpoint.

No solutions, just thought I could provide a little more context :-)

**Comment by rowanwins** [comment]
I _think_ geotiff.js would work fine with presigned urls as the auth info is sent via url params rather than headers 🤞

**Comment by DanielJDufour** [comment]
Agreed 👍  I'll keep my fingers crossed, too. haha

**Comment by jamesfisher-gis** [comment]
Hi there,
Has anyone considered this further? I am interested in implementing browsing secure imagery but could use some help. 

I am starting out by trying to have secure thumbnail assets to display in the browser. I have gotten secure thumbnails to display in a single-item view, but am struggling to implement my solution in the collection and catalog view. 

Here is my approach:
1. Create a token server with AWS Lambda & Gateway API. The token server receives a bucket and key, then returns the following:
```
{
"presigned_url": <signed URL>
}
```

2. In ThumbnailCardMixin.js, modify the thumbnail() computed method. Assign the signed URL to the returned thumbnail src prop.
```
thumbnail() {
      if (this.data) {
        let thumbnails = this.data.getThumbnails(true, 'thumbnail');
        if (thumbnails.length > 0) {
          let t = thumbnails[0];
          let width, height;
          if (Array.isArray(t['proj:shape']) && t['proj:shape'].length === 2) {
            [height, width] = t['proj:shape'];
          }
          else if (Array.isArray(this.defaultThumbnailSize) && this.defaultThumbnailSize.length === 2) {
            [height, width] = this.defaultThumbnailSize;
          }
    
          // eslint-disable-next-line;
          const regex = /\/\/([^\/]+)\/(.+)/;
            
          const matches = t.href.match(regex);
      
          // eslint-disable-next-line;
          const key = matches[2].replace(/^\//, '');
    
          // Generate signed URL using a token server
          axios.get(`<token server url>?key=${key}`).then(response => {
            const presigned_url = response.data.presigned_url
          console.log(presigned_url)
            // Return the thumbnail object with the signed URL
            return {
              src: presigned_url,
              alt: t.title,
              crossorigin: this.crossOriginMedia,
              right: this.isList,
              blankColor: "rgba(0, 0, 0, 0.125)",
              width,
              height
            };
          });
        }
      }
      return null;
    }
```

I see that the browser successfully makes requests to the token server, but it still requests the unsigned urls when displaying the thumbnails.

Is this sound like the right approach? I think this would be a solution the token server endpoint can be specified in the browser config.

**Comment by m-mohr** [comment]
See also https://github.com/radiantearth/stac-browser/issues/499#issue-2632988690

**Comment by m-mohr** [comment]
See also https://github.com/radiantearth/stac-browser/issues/372#issuecomment-1741708982
